cirq-core/cirq/transformers/transformer_primitives.py

```diff
@@ -304,54 +312,224 @@ class _MergedCircuit:
+                (this is necessary as component merges alter the component data)
+        """
+        self.components_by_index[c_data.moment].pop(c)
+        for q in c_data.qubits:
+            self.remove_moment(self.qubit_indexes[q], c_data.moment)
+        for mkey in c_data.mkeys:
+            self.remove_moment(self.mkey_indexes[mkey], c_data.moment)
+        for ckey in c_data.ckeys:
+            self.remove_moment(self.ckey_indexes[ckey], c_data.moment)
+
+    def get_mergeable_components(self, c: Component, c_qs: set[cirq.Qid]) -> list[Component]:
+        """Finds all components that can be merged with c.
+
+        Args:
+            c: component to be merged with existing components
+            c_qs: subset of c.qubits used to decide which components are mergeable
+
+        Returns:
+            list of mergeable components
+        """
+        # Find the index of previous moment which can be merged with `c`.
+        idx = max([self.qubit_indexes[q][-1] for q in c_qs], default=-1)
+        idx = max([idx] + [self.mkey_indexes[ckey][-1] for ckey in c.ckeys])
+        idx = max([idx] + [self.ckey_indexes[mkey][-1] for mkey in c.mkeys])
+        # Return the set of overlapping components in moment with index `idx`.
         if idx == -1:
-            return idx, []
+            return []
+
+        return [c for c in self.components_by_index[idx] if not c_qs.isdisjoint(c.qubits)]
+
+    def get_cirq_circuit(
+        self, components: list[Component], merged_circuit_op_tag: str
+    ) -> cirq.Circuit:
+        """Returns the merged circuit.
+
+        Args:
+            components: all components in creation order
+            merged_circuit_op_tag: tag to use for CircuitOperations
+
+        Returns:
+            the circuit with merged components as a CircuitOperation
+        """
+        component_ops: dict[Component, list[cirq.Operation]] = defaultdict(list)
+
+        # Traverse the components in creation order and collect operations
+        for c in components:
+            root = c.find()
+            component_ops[root].append(c.op)
+
+        moments = []
+        for m in self.components_by_index:
+            ops = []
+            for c in m.keys():
+                if isinstance(c, ComponentWithCircuitOp):
+                    ops.append(c.circuit_op)
+                    continue
+                if len(component_ops[c]) == 1:
+                    ops.append(component_ops[c][0])
+                else:
+                    ops.append(
+                        circuits.CircuitOperation(
+                            circuits.FrozenCircuit(component_ops[c])
+                        ).with_tags(merged_circuit_op_tag)
+                    )
+            moments.append(circuits.Moment(ops))
+        return circuits.Circuit(moments)
+
+
+def _merge_operations_impl(
+    circuit: CIRCUIT_TYPE,
+    factory: ComponentFactory,
+    *,
+    merged_circuit_op_tag: str = "Merged connected component",
```

Comment:
Is it a good practice to put space in a tag? @NoureldinYosri  I am not sure, though existing code before this pr have some mixed style of tags.

maybe "merged_connected_component"? or "_merged_connected_component" (start with _ to separate `a default tag named by cirq pkg` with `user specified tag`)?

Reply:
I copied the tag value from line 621 in merge_operations_to_circuit_op. I think the two should be in sync.